src/js/components/OptionalEnviromentComponent.jsx

```diff
@@ -0,0 +1,102 @@
+var React = require("react/addons");
+
+var appValidator = require("../validators/appValidator");
+var FormGroupComponent = require("../components/FormGroupComponent");
+var DuplicableRowControls = require("../components/DuplicableRowControls");
+
+var OptionalSettingsComponent = React.createClass({
+  displayName: "OptionalEnvironmentComponent",
+
+  propTypes: {
+    errors: React.PropTypes.array,
+    model: React.PropTypes.object.isRequired
+  },
+
+  getInitialState: function () {
+    var state = {
+      env: Object.keys(this.props.model.env).map(function (key) {
+        return {
+          key: key,
+          value: this.props.model.env[key]
+        };
+      })
+    };
+    if (state.env.length === 0) {
+      state.env = [true];
+    }
+    return state;
+  },
+
+  handleAddRow: function (rowKey, position, event) {
+    event.target.blur();
+    event.preventDefault();
+    var state = this.state;
+    state.env.push(true); // = [true];
+    this.setState(state);
+  },
+
+  handleRemoveRow: function (rowKey, position, event) {
+    /* Each array represents a list of duplicable rows, in order to keep track
+     * of which rows have been deleted by the user. A value of false signifies
+     * a deleted row.
+     * The form submits values infered directly from the DOM.
+     */
+    event.target.blur();
+    event.preventDefault();
+    var state = {
```

Comment:
I would write it more extensible:
```
var env = this.state.env.filter(function (value, index) {
   return index !== position;
});

this.setState({
  env: env
});
```

setState only overrides the properties you are setting. Other existing properties are preserved.